lib/collection/src/shards/replica_set.rs

```diff
@@ -687,6 +688,114 @@ impl ShardReplicaSet {
         captured_error.expect("at this point `captured_error` must be defined by construction")
     }
 
+    pub async fn execute_and_resolve_read_operation<'a, F, Fut, Res>(
+        &self,
+        read_operation: F,
+        local: &'a Option<Shard>,
+        remotes: &'a [RemoteShard],
+        factor: usize,
+        condition: ResolveCondition,
+    ) -> CollectionResult<Res>
+    where
+        F: Fn(&'a (dyn ShardOperation + Send + Sync)) -> Fut,
+        Fut: Future<Output = CollectionResult<Res>>,
+        Res: Resolve,
+    {
+        let local_count = usize::from(local.is_some());
+        let remotes_count = remotes.len();
+
+        let active_local = local
+            .as_ref()
+            .filter(|_| self.peer_is_active(&self.this_peer_id()));
+
+        let active_remotes_iter = remotes
+            .iter()
+            .filter(|remote| self.peer_is_active(&remote.peer_id));
+
+        let active_local_count = usize::from(active_local.is_some());
+        let active_remotes_count = active_remotes_iter.clone().count();
+
+        let total_count = local_count + remotes_count;
+        let active_count = active_local_count + active_remotes_count;
+
+        let factor = factor.clamp(1, total_count);
+
+        if active_count < factor {
+            return Err(CollectionError::service_error(format!(
+                "The replica set for shard {} on peer {} does not have enough active replicas",
+                self.shard_id,
+                self.this_peer_id(),
+            )));
+        }
+
+        let mut active_remotes: Vec<_> = active_remotes_iter.collect();
+        active_remotes.shuffle(&mut rand::thread_rng());
+
+        let local_operations = active_local
+            .into_iter()
+            .map(|local| read_operation(local.get()).left_future());
+
+        let remote_operations = active_remotes
+            .into_iter()
+            .map(|remote| read_operation(remote).right_future());
+
+        let mut operations = local_operations.chain(remote_operations);
+
+        let mut pending_operations: FuturesUnordered<_> = operations
+            .by_ref()
+            .take(factor + usize::try_from(self.read_remote_replicas).unwrap())
+            .collect();
+
+        let mut responses = Vec::new();
+
+        while let Some(result) = pending_operations.next().await {
+            match result {
+                Ok(resp) => responses.push(resp),
+
+                Err(err) => {
+                    let is_transient = matches!(
+                        &err,
+                        CollectionError::ServiceError { .. } | CollectionError::Cancelled { .. },
+                    );
+
+                    if is_transient {
+                        log::debug!("Read operation failed: {err}");
+                    } else {
+                        log::error!("Read operation failed: {err}");
```

Comment:
Don't need this log, it will be logged later

Reply:
Removed.